### PR TITLE
chore(telemetry): track agent connection configurations

### PR DESCRIPTION
**Pull request by mabdinur**
## Description

- Ensures datadog trace agent environment variables are only read on application start up. After application start up these configurations should only be updated via remote configuration. Remote Configuration support is currently in development.
  - The following environment variables have been updated:
    - DD_DOGSTATSD_PORT
    - DD_DOGSTATSD_URL
    - DD_AGENT_HOST
    - DD_AGENT_PORT
    - DD_TRACE_AGENT_URL
    - DD_TRACE_AGENT_TIMEOUT_SECONDS
- Converts agent connection tests to subprocess tests. Since environment variables are only read when `ddtrace.config` is initialized this is required for some tests to pass. 
- Replaces all references to `get_trace_hostname`, `get_stats_hostname`, `get_trace_port`, `get_stats_port`, `get_trace_agent_timeout` with `config._trace_agent_hostname`, `config._stats_agent_hostname`, `config._trace_agent_port`, `config._stats_agent_port`, and  `config._agent_timeout_seconds`.
- Sends agent configurations to the instrumentation telemetry platform.

## Motivation
- Track agent configurations via the Instrumentation Telemetry platform

## Checklist

- [x] Change(s) are motivated and described in the PR description.
- [x] Testing strategy is described if automated tests are not included in the PR.
- [x] Risk is outlined (performance impact, potential for breakage, maintainability, etc).
- [x] Change is maintainable (easy to change, telemetry, documentation).
- [x] [Library release note guidelines](https://ddtrace.readthedocs.io/en/stable/releasenotes.html) are followed. If no release note is required, add label `changelog/no-changelog`.
- [x] Documentation is included (in-code, generated user docs, [public corp docs](https://github.com/DataDog/documentation/)).
- [x] Backport labels are set (if [applicable](https://ddtrace.readthedocs.io/en/latest/contributing.html#backporting))

## Reviewer Checklist

- [x] Title is accurate.
- [x] No unnecessary changes are introduced.
- [x] Description motivates each change.
- [x] Avoids breaking [API](https://ddtrace.readthedocs.io/en/stable/versioning.html#interfaces) changes unless absolutely necessary.
- [x] Testing strategy adequately addresses listed risk(s).
- [x] Change is maintainable (easy to change, telemetry, documentation).
- [x] Release note makes sense to a user of the library.
- [x] Reviewer has explicitly acknowledged and discussed the performance implications of this PR as reported in the benchmarks PR comment.
- [x] Backport labels are set in a manner that is consistent with the [release branch maintenance policy](https://ddtrace.readthedocs.io/en/latest/contributing.html#backporting)
- [x] If this PR touches code that signs or publishes builds or packages, or handles credentials of any kind, I've requested a review from `@DataDog/security-design-and-guidance`.
- [x] This PR doesn't touch any of that.

### CODE REVIEW

File: ddtrace/internal/agent.py [hunk 1]
```diff
@@ -41,81 +39,43 @@ def is_ipv6_hostname(hostname):
...
-    if not url:
-        if user_supplied_host or user_supplied_port:
-            url = "http://%s:%s" % (get_trace_hostname(), get_trace_port())
-        elif os.path.exists("/var/run/datadog/apm.socket"):
-            url = "unix://%s" % (DEFAULT_UNIX_TRACE_PATH)
-        else:
-            url = DEFAULT_TRACE_URL
-
-    return url
+    if ddconfig._trace_agent_url is not None:
```
Comment (mabdinur) [inline ddtrace/internal/agent.py:48]:
```suggestion
    if ddconfig._trace_agent_url:
```

File: ddtrace/internal/agent.py [hunk 2]
```diff
@@ -41,81 +39,43 @@ def is_ipv6_hostname(hostname):
...
-
-    if not url:
-        if user_supplied_host or user_supplied_port:
-            url = "udp://{}:{}".format(get_stats_hostname(), get_stats_port())
-        elif os.path.exists("/var/run/datadog/dsd.socket"):
-            url = "unix://%s" % (DEFAULT_UNIX_DSD_PATH)
-        else:
-            url = "udp://{}:{}".format(get_stats_hostname(), get_stats_port())
-    return url
+    if ddconfig._stats_agent_url is not None:
```
Comment (mabdinur) [inline ddtrace/internal/agent.py:63]:
```suggestion
    if ddconfig._stats_agent_url:
```

File: ddtrace/settings/config.py [hunk 3]
```diff
@@ -270,6 +271,15 @@ def __init__(self):
         )
         self._trace_writer_log_err_payload = asbool(os.environ.get("_DD_TRACE_WRITER_LOG_ERROR_PAYLOADS", False))
 
+        self._trace_agent_hostname = os.environ.get("DD_AGENT_HOST", os.environ.get("DD_TRACE_AGENT_HOSTNAME"))
```
[the remark this answers is not in the document]
Comment (mabdinur) [reply ddtrace/settings/config.py:274]: I am planning to do this in a follow up PR.  Making this change would involve updating every reference to `get_trace_url()` and `get_stats_url()` . This change would impact over 11 files and require approvals from ASM, debugger, profiling and ci-visibility.

[the remark this answers is not in the document]
Comment (ZStriker19) [reply ddtrace/settings/config.py:274]: understood!

[the remark this answers is not in the document]
Comment (mabdinur) [reply ddtrace/settings/config.py:274]: I tried to keep the changes in this PR as self contained as possible.

**Review by ZStriker19 — DISMISSED** [review]

**Review by ZStriker19 — APPROVED** [review]

**Review by ZStriker19 — APPROVED** [review]

**Review by ZStriker19 — DISMISSED** [review]

**Review by ZStriker19 — APPROVED** [review]

**Pull request merged**